lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -832,5 +920,25 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       return new DefaultBulkScorer(new ConstantScoreScorer(this, 0f, ScoreMode.COMPLETE_NO_SCORES, disi));
     }
 
+    // Perform a cache load asynchronously
+    private void cacheAsynchronously(LeafReaderContext context, IndexReader.CacheHelper cacheHelper) throws RejectedExecutionException {
+      /*
+       * If the current query is already being asynchronously cached,
+       * do not trigger another cache operation
+       */
+      if (inFlightAsyncLoadQueries.add(in.getQuery()) == false) {
+        return;
+      }
+
+      FutureTask<Void> task = new FutureTask<>(() -> {
+        DocIdSet localDocIdSet = cache(context);
+        putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+
+        //Remove the key from inflight -- the key is loaded now
+        inFlightAsyncLoadQueries.remove(in.getQuery());
```

Comment:
Can you `assert` the return value here?  I.e. we should *always* succeed in removing this query from the set.